### DISCUSSION TITLE
Injecting Z axis movement manually while running GCode

### DISCUSSION THREAD

**Discussion by fireup924** [grblHAL/RP2040#76] [Ideas]
I have a machine that is basically a 2 axis machine with the Z axis not moving while running a GCode. The Z does need to be adjusted manually while the program is running using UP/Down buttons. How can I incorporate this feature?

**Comment by terjeio** [comment]
You'll have to write a plugin that uses the new step inject functionality. Easiest will be to call the [new stepper functions](https://github.com/grblHAL/core/tree/04a5edcf8353f7294167d12d32e4b073d9be9ce4) that sits on top of  of step inject functionality as these handles acceleration/deceleration etc.

FYI I added supporting low-level code to this driver a few days ago but has not published it yet since I do not have means to test it until I am back home in early December. The new stepper functions has also been changed somewhat in yet to be published code, testing beeing done with the STM32F4xx driver.

Are you able to write and test plugin code yourself? I can provide some guidance.

**Reply by fireup924** [reply]
> Are you able to write and test plugin code yourself? I can provide some guidance.

Yes I can modify a plugin and test it myself.

I see that there's flag STEP_INJECT_ENABLE in iMXRT1062 and STM32F4xx but not in RP2040. Will this work for RP2040 also?

**Reply by terjeio** [reply]
> Will this work for RP2040 also?

No, not with the code currently on gihub. I can give you read access to the master subversion repo for downloading the latest bleeding edge version. Post an email address in a comment then edit it afterwards and remove it and I'll send you information about how to. Also specify wheter you are compiling on Windows or linux.

**Reply by fireup924** [reply]
Thanks for the link to the latest code.

I was able to get the Plasma plugin to compiled for RP2040. This plugin looks to be very close to what I need. Where should I insert an input reading in thc.c and use the two inputs for UP/Down instead of arc_voltage?

**Reply by terjeio** [reply]
If the board map has 3 or more aux inputs available then set the mode ($350) to 2 enables up/down mode. But note this mode is currently not using acceleration/deceleration of the motor - not sure how to implement that... It might be better for you to start a rather long motion on key down events and stop it on key up events? Kind of like continuous jogging.

**Reply by fireup924** [reply]
I enabled (#350) to 2, but it does not seem to be working. How can I get the status of Aux inputs so I know it's working? What should ($367) be set to? What command do I need to send to get a |THC: report?

**Reply by terjeio** [reply]
Did you compile with the version from the subversion repo?

> How can I get the status of Aux inputs so I know it's working?

Compile without the plasma plugin and use `M66PxL0` where x is the aux input number. Then request a status report to get the status, e.g:

<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,0|Pn:PXYZ|**In:1**>

Use `$pins` to list the pin mappings. BTW when the plasma plugin successfully claims the needed inputs then it will show in the `$pins` listing. "THC" will also show up in the NEWOPT list returned by `$I`:

[OPT:VNMSL,35,1024,3,0]
[AXS:3:XYZ]
[NEWOPT:ENUMS,RT+,ES,SED,RTC,**THC**]
[FIRMWARE:grblHAL]
[NVS STORAGE:*FLASH]
[FREE MEMORY:109K]
[DRIVER:STM32F446]
[DRIVER VERSION:231021]
[BOARD:Nucleo-64 CNC Breakout]
[AUX IO:1,7,0,3]
[PLUGIN:PLASMA v0.10]

Here are parts of the `$pins` listing when the plugin has claimed the ports:
```
[PIN:PB14,Aux input 0,P0]
[PIN:PA15,Aux input 1,Cutter up]
[PIN:PB13,Aux input 2,Cutter down]
[PIN:PB0,Aux input 3,Arc ok]
[PIN:PC0,X step]
...
[PIN:PA8,Spindle PWM]
[PIN:MCP3221:0,Aux analog input 0,Arc voltage]
```

> What should ($367) be set to?

The port number (Pn in the $pins listing) used for the ARC ok input.
`$368` and `$369` is for binding cutter down and up ports.
You may try resetting plugin settings to default with `$RST=&` to get sensible default values.

![image](https://github.com/grblHAL/RP2040/assets/20260062/d742eb6c-1de5-491d-b1ec-bab345c74f18)

> What command do I need to send to get a |THC: report?

A realtime status request: `?`
Note is it only available when the plugin successfuly claims the required inputs on startup.

Finally: the up/down inputs will not be working unless there is XY motion and the `Arc ok` signal is detected as asserted - which requires the spindle to be running.

**Reply by fireup924** [reply]
Thank you for all the things I can try

Yes I compiled with the the subversion repo.

I was able to get inputs to report it's status and confirm it's working. and sending $pins return

[PIN:13,Aux input 0,Arc ok]
[PIN:14,Aux input 1,Cutter down]
[PIN:15,Aux input 2,Cutter up]

I had to do settings default reset `$RST=& ` so I can assign $368 and $369 properly.

I'm now able to get the |THC:AETU and |THC:AETD status to indicate that the Up or Down inputs are active, but there's no movement in the Z?

**Reply by terjeio** [reply]
> ... but there's no movement in the Z?

It works for me with the STM32F4xx board.
Is the Z stepper enabled? If not try setting $37 to 4, this leaves the Z stepper always enabled. If it still does not work then step injection is not working on the RP2040 or the step rate is too high (fixed at 1KHz for up/down mode for now - no acceleration/deceleration).

Which board map are you using?

**Reply by fireup924** [reply]
Ok, the DRO for Z is not changing but was able to see the step signal on a scope (this is fine if that's how it should work). The DIR signal is not changing for me for UP or Down input. (Z does works in normal Jog Up/Down)

> Which board map are you using?

I'm using a custom board that is very similar to pico_cnc that has 74HTC595 for step/dir

How can I change the thc.c code so it uses the Jog Z up/down instead of step_inject function? This will make it easier for me to change to speed and acceleration.

**Reply by fireup924** [reply]
Ok I found the bug for the DIR signal not working. It's in the driver.c on line 1036, (the if statement was closed too early ) and the end of
"if (step_outbits.value)" condition needs needs to be closed on line 1072.

and also I need to #define DIRECTION_OUTMODE  GPIO_SR8 in my board map

**Reply by terjeio** [reply]
> Ok, the DRO for Z is not changing but was able to see the step signal on a scope (this is fine if that's how it should work). 

This is per design - step injection for tool height control should not change the DRO position. You'll have to add the step count to `sys.position[Z_AXIS]` when you need to update it.

> How can I change the thc.c code so it uses the Jog Z up/down instead of step_inject function?

You can send jog commands to the parser only when the controller is in idle (or jog) state. When motion is ongoing you will have to use step injection. Acceleration/deceleration is possible if you use _stepper2.c_ functions instead of injecting steps directly - voltage controlled THC does that. 

I'll fix the bugs/typos in driver.c - by reorganizing the code a bit.

**Reply by fireup924** [reply]
>  Acceleration/deceleration is possible if you use stepper2.c functions instead of injecting steps directly - voltage controlled THC does that.

So can I fake a voltage by converting a button press to a value (Up press -> arc_voltage = 250, Down press -> arc_voltage = 5)? and it will get me Acceleration/deceleration?

**Reply by terjeio** [reply]
Voltage mode uses feedback to keep the voltage, and thus the position above the material, constant. Faking the voltage to achieve what you want might be pretty hard...
As I wrote earlier it is better to detect both when a key is pressed and released, on pressed send a relatively large move and on key released stop it by calling `st2_motor_stop()`.

FYI have added a bit of documentation to _stepper2.c_ and _stepper2.h_ in the latest commit to the subversion repo.

**Reply by fireup924** [reply]
Ok I understand now.  

When I see a state change in the Up/Down inputs, I just use st2_motor_move(), st2_motor_stop() and st2_motor_run() directly.

**Reply by fireup924** [reply]
Thank you for all you help, everything is working the way I like now! :)

**Reply by terjeio** [reply]
:+1:

**Comment by fireup924** [comment]


**Reply by terjeio** [reply]
Email sent.
Code that uses step injection via `grbl/stepper2.c` functions can be found in the [plasma](https://github.com/grblHAL/Plugin_plasma/blob/master/thc.c) and [spindle](https://github.com/grblHAL/Plugins_spindle/blob/master/stepper.c) plugins. Note: both of these has been updated, refer to the copies in the subversion repo.
